drivers/i2c/i2c_omap.c

```diff
@@ -0,0 +1,710 @@
+ * i2c_omap_recover_bus().
+ *
+ * @param dev The I2C device structure.
+ * @return 0 if the bus becomes free, or a negative error code if the bus cannot
+ * be recovered.
+ */
+static int i2c_omap_wait_for_bb(const struct device *dev)
+{
+	i2c_omap_regs_t *i2c_base_addr = DEV_I2C_BASE(dev);
+	uint32_t timeout = k_uptime_get_32() + I2C_OMAP_TIMEOUT;
+
+	while (i2c_base_addr->STAT & I2C_OMAP_STAT_BB) {
+		if (k_uptime_get_32() > timeout) {
+			LOG_ERR("Bus busy timeout");
+#ifdef CONFIG_I2C_OMAP_BUS_RECOVERY
+			return i2c_omap_recover_bus(dev);
+#else
+			return -ETIMEDOUT;
+#endif /* CONFIG_I2C_OMAP_BUS_RECOVERY */
+		}
+		k_busy_wait(100);
+	}
+	return 0;
+}
+
+/**
+ * @brief Performs data transfer for the OMAP I2C driver.
+ *
+ * This function is responsible for handling the data transfer logic for the OMAP I2C driver.
+ * It reads the status register and performs the necessary actions based on the status flags.
+ * It handles both receive and transmit logic, and also handles error conditions such as NACK,
+ * arbitration lost, receive overrun, and transmit underflow.
+ *
+ * @param dev Pointer to the device structure.
+ *
+ * @return Returns 0 on success, or a negative error code on failure.
+ */
+static int i2c_omap_transfer_message_ll(const struct device *dev)
+{
+	struct i2c_omap_data *data = DEV_DATA(dev);
+	i2c_omap_regs_t *i2c_base_addr = DEV_I2C_BASE(dev);
+	uint16_t stat = i2c_base_addr->STAT, result = 0;
+
+	if (data->receiver) {
+		stat &= ~(I2C_OMAP_STAT_XDR | I2C_OMAP_STAT_XRDY);
+	} else {
+		stat &= ~(I2C_OMAP_STAT_RDR | I2C_OMAP_STAT_RRDY);
+	}
+	if (stat & I2C_OMAP_STAT_NACK) {
+		result |= I2C_OMAP_STAT_NACK;
+		i2c_base_addr->STAT |= I2C_OMAP_STAT_NACK;
+	}
+	if (stat & I2C_OMAP_STAT_AL) {
+		result |= I2C_OMAP_STAT_AL;
+		i2c_base_addr->STAT |= I2C_OMAP_STAT_AL;
+	}
+	if (stat & I2C_OMAP_STAT_ARDY) {
+		i2c_base_addr->STAT |= I2C_OMAP_STAT_ARDY;
+	}
+	if (stat & (I2C_OMAP_STAT_ARDY | I2C_OMAP_STAT_NACK | I2C_OMAP_STAT_AL)) {
+
+		i2c_base_addr->STAT |=
+			(I2C_OMAP_STAT_RRDY | I2C_OMAP_STAT_RDR | I2C_OMAP_STAT_XRDY |
+			 I2C_OMAP_STAT_XDR | I2C_OMAP_STAT_ARDY);
+		return result;
+	}
+
+	/* Handle receive logic */
+	if (stat & (I2C_OMAP_STAT_RRDY | I2C_OMAP_STAT_RDR)) {
+		int buffer =
+			(stat & I2C_OMAP_STAT_RRDY) ? i2c_base_addr->BUF : i2c_base_addr->BUFSTAT;
+		i2c_omap_transmit_receive_data(dev, buffer);
+		i2c_base_addr->STAT |=
+			(stat & I2C_OMAP_STAT_RRDY) ? I2C_OMAP_STAT_RRDY : I2C_OMAP_STAT_RDR;
+		return RETRY;
+	}
+
+	/* Handle transmit logic */
+	if (stat & (I2C_OMAP_STAT_XRDY | I2C_OMAP_STAT_XDR)) {
+		int buffer =
+			(stat & I2C_OMAP_STAT_XRDY) ? i2c_base_addr->BUF : i2c_base_addr->BUFSTAT;
+		i2c_omap_transmit_receive_data(dev, buffer);
+		i2c_base_addr->STAT |=
+			(stat & I2C_OMAP_STAT_XRDY) ? I2C_OMAP_STAT_XRDY : I2C_OMAP_STAT_XDR;
+		return RETRY;
+	}
+
+	if (stat & I2C_OMAP_STAT_ROVR) {
+		result |= I2C_OMAP_STAT_ROVR;
```

Comment:
You could just return `I2C_OMAP_STAT_ROVR` directly, no need to add to `result` just to return that in the next line.

Reply:
Removed it

Reply:
Sorry for that.
I have updated it now.